appserver/web/war-util/src/main/java/org/glassfish/web/loader/WebappClassLoader.java

```diff
@@ -837,56 +838,58 @@ public synchronized void addJar(String jar, JarFile jarFile, File file)
             logger.log(Level.FINER, "addJar({0})", jar);
         }
 
-        // See IT 11417
-        super.addURL(getURL(file));
+        synchronized (jarFilesPropertyLock) {
```

Comment:
The method is already synchronized, how could this help? Two locks can cause deadlock ... but ... in previous version jarFiles are used also as a lock, so that's the same ...

Reply:
the method addJar() is synchronized, but other parts of the class requires synchronized access to the property jarFiles.  But the synchronization on the method does not synchronize on the property.

In this class, there are synchronization block on 6 different items.